pageserver/src/thread_mgr.rs

```diff
@@ -0,0 +1,297 @@
+}
+
+/// Launch a new thread
+pub fn spawn<F, E>(
+    kind: ThreadKind,
+    tenant_id: Option<ZTenantId>,
+    timeline_id: Option<ZTimelineId>,
+    name: &str,
+    f: F,
+) -> std::io::Result<()>
+where
+    F: FnOnce() -> Result<(), E>,
+    F: Send + 'static,
+{
+    let (shutdown_tx, shutdown_rx) = watch::channel(());
+    let thread_id = NEXT_THREAD_ID.fetch_add(1, Ordering::Relaxed);
+    let thread = PageServerThread {
+        _thread_id: thread_id,
+        kind,
+        name: name.to_string(),
+
+        shutdown_requested: AtomicBool::new(false),
+        shutdown_tx,
+
+        inner: Mutex::new(PageServerThreadInner {
+            tenant_id,
+            timeline_id,
+            join_handle: None,
+        }),
+    };
+
+    let thread_rc = Arc::new(thread);
+
+    let mut inner_guard = thread_rc.inner.lock().unwrap();
+
+    THREADS
+        .lock()
+        .unwrap()
+        .insert(thread_id, Arc::clone(&thread_rc));
+
+    let thread_rc2 = Arc::clone(&thread_rc);
+    let join_handle = match thread::Builder::new()
+        .name(name.to_string())
+        .spawn(move || thread_wrapper(thread_id, thread_rc2, shutdown_rx, f))
+    {
+        Ok(handle) => handle,
+        Err(err) => {
+            // Could not spawn the thread. Remove the entry
+            THREADS.lock().unwrap().remove(&thread_id);
+            return Err(err);
+        }
+    };
+    inner_guard.join_handle = Some(join_handle);
+    drop(inner_guard);
+
+    // The thread is now running. Nothing more to do here
+    Ok(())
+}
+
+/// This wrapper function runs in a newly-spawned thread. It initialies the thread-local
+/// variables and calls the payload function
+fn thread_wrapper<F, E>(
+    thread_id: u64,
+    thread: Arc<PageServerThread>,
+    shutdown_rx: watch::Receiver<()>,
+    f: F,
+) where
+    F: FnOnce() -> Result<(), E>,
+    F: Send + 'static,
+{
+    SHUTDOWN_RX.with(|rx| {
+        *rx.borrow_mut() = Some(shutdown_rx);
+    });
+    CURRENT_THREAD.with(|ct| {
+        *ct.borrow_mut() = Some(thread);
+    });
+
+    // FIXME: explain why AssertUnwindSafe is ok
+    let result = panic::catch_unwind(AssertUnwindSafe(f));
+
+    THREADS.lock().unwrap().remove(&thread_id);
+
+    if let Err(err) = result {
+        panic::resume_unwind(err);
+    }
+}
+
+/// Is there a thread running that matches the criteria
+///
+/// Any None arguments are ignored. For example, to check if there are any WalReceiver
+/// threads running:
+///
+///   is_thread_running(Some(ThreadKind::WalReceiver), None, None)
+///
+/// Or to check if there are any threads for given timeline runnning:
+///
+///   is_thread_running(None, Some(timelineid), None)
+///
+pub fn is_thread_running(
+    kind: Option<ThreadKind>,
+    tenant_id: Option<ZTenantId>,
+    timeline_id: Option<ZTimelineId>,
+) -> bool {
+    let threads = THREADS.lock().unwrap();
+    for thread in threads.values() {
+        if kind.is_none() || thread.kind == kind.unwrap() {
```

Comment:
NIT: That unused function is a bit odd IMO, the `unwrap` could be rewritten in more comprehensible way, also, `is_none` path does not really require any locking, which might be nice too.
IMO, `match` is enough to handle those in a readable manner.

Similar idea for the shutdown function.

Reply:
> NIT: That unused function is a bit odd

Ah, didn't realize it was unused. I had a use case for it in mind at some point, but don't remember what it was. Removed.

> IMO, the unwrap could be rewritten in more comprehensible way, also, is_none path does not really require any locking, which might be nice too.
IMO, match is enough to handle those in a readable manner.

Can you help me write that with `match`?

Reply:
Hm, either I've got wrong, or there were two locks here in some verison, now that's gone and not that much can be really done.

I'd replaced `thread.kind == kind.unwrap()` with `kind == Some(thread.kind)`, but that's not necessary.